docs/flakty-tests.md

```diff
@@ -0,0 +1,260 @@
+    break;
+  } catch(e) {
+    // if there was an error, check if we've already been trying for too long
+    if((new Date() - startTime) > MAX_WAIT_TIME) {
+      throw e;
+    } else {
+      // wait a little while
+      await sleep(1);
+    }
+  }
+}
+```
+
+Of course, our interactors perform this transformation automatically, and you
+can just use our easy to use API and rest assured that we have automatically
+tranformed this into the above form for you! Just write it like this:
+
+``` javascript
+await Form.byId('login').find(TextField.byId('name')).fillIn('Jonas');
+```
+
+There are of course a number of details that are omitted in the simplified
+code above. For example, each interactor must be specific enough to match a
+*single* element on the page. This means that if there were multiple fields
+with the id `name` then the above interaction would fail. This forces you to
+be specific enough so that you don't end up interacting with a random element
+on the page, and your tests might break due to layout changes where the order
+of elements changes.
+
+### Taming the event loop
+
+One huge advantage that we have over Capybara is that since we are running tests
+directly in the browser, we can leverage the event loop, and ensure that our
+interactions run fully within the same tick of the loop. This eliminates a lot
+of problems which plague Capybara and similar frameworks based on Selenium and
+similar tools. In Capybara we go to great lengths to avoid problems with the
+dreaded `StaleReferenceError`, but we simply do not need to do this in BigTest.
+By running within the event loop, and ensuring that we never release it, we can
+ensure that our elements are never stale.
+
+This has some implications on the way that interactors are defined, but we
+have worked very hard to make it as intuitive as possible to write
+interactors in such a way that they will be very resistant to problems on
+this front. Of course, BigTest will also ship with a rich library of
+interactors already built in, for all of the most common interface elements
+in web applications, so most users never have to define their own interactors
+if they don't want to.
+
+The TextField interactor we used above could be defined like this:
+
+``` typescript
+const TextField = createInteractor<HTMLInputElement>('text field')({
+  selector: 'input',
+  locators: {
+    byId: (element) => element.id,
+  },
+  actions: {
+    fillIn: (element, value: string) => { element.value = value }
+  }
+});
+```
+
+Both the locator and action functions are synchronous functions, which means
+that they do not release the event loop. This allows us to transform them
+into convergent functions on the interactor which perform the entire
+interaction of finding the element and runnign the functions against them.
+
+BigTest is written in TypeScript and our interactors are fully typesafe when
+used from TypeScript both in their definition and when using them.
+
+## Inherent ambiguities
```

Comment:
Perhaps we can consider this?

_Further challenges: Inherent ambiguities_

Reply:
I have changed all three of the big headlines into the format `Challenge: xyz`, I think this gives it some more easily digestible structure, let me know what you think!